File: dotnet/src/Envdoctor/Commands/ScanCommand.cs
using Envdoctor.Core;
using Envdoctor.Detectors;
using Envdoctor.Formatters;
using Envdoctor.Models;
using Envdoctor.Utils;

namespace Envdoctor.Commands;

public sealed class ScanArgs
{
    public OutputArgs Output { get; set; } = new();
    public string? Root { get; set; }
    public bool Verbose { get; set; }
    public List<string> Only { get; set; } = new();
    public string? Baseline { get; set; }
    public string? WriteBaseline { get; set; }
    public bool Staged { get; set; }
    public string? Since { get; set; }
    public bool Json { get; set; }
}

public static class ScanCommand
{
    public static int Run(ScanArgs args)
    {
        var root = Path.GetFullPath(args.Root ?? ".");

        // Warn about unknown detector ids passed via --only.
        var known = Audit.AllDetectors().Select(d => d.Id).ToHashSet();
        foreach (var rule in args.Only)
        {
            if (!known.Contains(rule))
            {
                var ids = known.ToList();
                ids.Sort(StringComparer.Ordinal);
                Console.Error.WriteLine($"warning Unknown detector \"{rule}\" (known: {string.Join(", ", ids)})");
            }
        }

        // Determine a git-changed file filter for --staged / --since.
        HashSet<string>? changed = args.Staged
            ? Discover.StagedFiles(root).ToHashSet()
            : args.Since is { } since
                ? Discover.ChangedFilesSince(root, since).ToHashSet()
                : null;
        var filterActive = changed is not null;

        var (model, config) = Pipeline.LoadProjectFiltered(root, changed);

        if (filterActive && model.AllFiles.Count == 0)
        {
            Console.Out.WriteLine("✓ No changed env-related files to scan");
            return 0;
        }

        var index = IndexedModel.BuildIndex(model);
        var findings = Audit.RunAudit(model, config, index);

        Shared.ReportParseErrors(model, root);

        // --only: restrict to the requested detector ids.
        if (args.Only.Count > 0)
        {
            var only = args.Only.ToHashSet();
            findings = findings.Where(f => only.Contains(f.RuleId)).ToList();
        }

        // --baseline: suppress previously-recorded findings.
        if (args.Baseline is { } baselinePath)
            findings = ApplyBaseline(findings, root, baselinePath);

        // --write-baseline: persist the current findings as a baseline.
        if (args.WriteBaseline is { } writePath)
            WriteBaselineFile(findings, root, writePath);

        var summary = Pipeline.Summarize(model, findings);

        // --json is an alias for --format json.
        var output = args.Output.Clone();
        if (args.Json)
            output.Format = OutputFormat.Json;

        // JSON output uses a stable, camelCase projection matching the
        // reference CLI (rather than the internal snake_case model shape).
        if (output.Format == OutputFormat.Json)
        {
            // Under --strict, warnings are promoted to errors for the exit code.
            var finalFindings = output.Strict
                ? findings.Select(f =>
                {
                    var clone = f.Clone();
                    if (clone.Severity == Severity.Warning)
                        clone.Severity = Severity.Error;
                    return clone;
                }).ToList()
                : findings;
            var exit = ExitCodes.AuditExitCode(new ExitContext { Findings = finalFindings, Strict = output.Strict });
            var json = RenderScanJson(root, finalFindings, summary, exit);
            if (output.Output is { } path)
                File.WriteAllText(path, json + "\n");
            else
                Console.Out.WriteLine(json);
            return exit;
        }

        return Shared.OutputFindingsVerbose(root, output, findings, summary, args.Verbose);
    }

    /// Render the scan result as the reference CLI's JSON shape.
    private static string RenderScanJson(string root, List<Finding> findings, AuditSummary summary, int exitCode)
    {
        var findingsJson = findings.Select(f =>
        {
            var locations = f.Locations.Select(o =>
            {
                var loc = new JsonObject { { "file", HumanFormatter.DisplayPath(root, o.FilePath) } };
                if (o.Line is { } line)
                    loc.Add("line", line);
                loc.Add("kind", o.Kind.AsStr());
                return (object)loc;
            }).ToList();
            return (object)new JsonObject
            {
                { "id", f.Id },
                { "ruleId", f.RuleId },
                { "severity", f.Severity.AsStr() },
                { "variable", f.Variable },
                { "message", f.Message },
                { "locations", locations },
            };
        }).ToList();

        return Json.Pretty(new JsonObject
        {
            { "exitCode", exitCode },
            {
                "summary", new JsonObject
                {
                    { "filesScanned", summary.FilesScanned },
                    { "variablesFound", summary.VariablesFound },
                    { "errors", summary.Errors },
                    { "warnings", summary.Warnings },
                    { "infos", summary.Infos },
                    { "total", summary.Total },
                }
            },
            { "findings", findingsJson },
        });
    }

    private sealed record BaselineEntry(string RuleId, string Variable, List<string> Files);

    private sealed class BaselineFile
    {
        public int Version { get; set; }
        public List<BaselineEntry> Findings { get; set; } = new();
    }

    private static BaselineEntry Fingerprint(string root, Finding finding)
    {
        var files = finding.Locations
            .Select(o => HumanFormatter.DisplayPath(root, o.FilePath))
            .Distinct()
            .ToList();
        files.Sort(StringComparer.Ordinal);
        return new BaselineEntry(finding.RuleId, finding.Variable, files);
    }

    private static bool EntryMatches(BaselineEntry a, BaselineEntry b) =>
        a.RuleId == b.RuleId && a.Variable == b.Variable && a.Files.SequenceEqual(b.Files);

    private static List<Finding> ApplyBaseline(List<Finding> findings, string root, string baselinePath)
    {
        var full = Path.Combine(root, baselinePath);
        BaselineFile? baseline = null;
        try
        {
            var raw = File.ReadAllText(full);
            using var doc = System.Text.Json.JsonDocument.Parse(raw);
            var rootEl = doc.RootElement;
            baseline = new BaselineFile
            {
                Version = rootEl.TryGetProperty("version", out var v) ? v.GetInt32() : 0,
                Findings = rootEl.TryGetProperty("findings", out var f) && f.ValueKind == System.Text.Json.JsonValueKind.Array
                    ? f.EnumerateArray().Select(e => new BaselineEntry(
                        e.GetProperty("rule_id").GetString() ?? "",
                        e.GetProperty("variable").GetString() ?? "",
                        e.GetProperty("files").EnumerateArray().Select(x => x.GetString() ?? "").ToList()))
                        .ToList()
                    : new List<BaselineEntry>(),
            };
        }
        catch
        {
            baseline = null;
        }

        if (baseline is null)
        {
            Console.Error.WriteLine($"warning Could not read baseline {baselinePath}");
            return findings;
        }

        var before = findings.Count;
        var kept = findings
            .Where(f => !baseline.Findings.Any(b => EntryMatches(b, Fingerprint(root, f))))
            .ToList();
        var suppressed = before - kept.Count;
        if (suppressed > 0)
        {
            Console.Error.WriteLine(
                $"info {suppressed} finding{(suppressed == 1 ? "" : "s")} suppressed by baseline");
        }
        return kept;
    }

    private static void WriteBaselineFile(List<Finding> findings, string root, string baselinePath)
    {
        var full = Path.Combine(root, baselinePath);
        var baseline = new JsonObject
        {
            { "version", 1 },
            {
                "findings", findings
                    .Select(f =>
                    {
                        var fp = Fingerprint(root, f);
                        return (object)new JsonObject
                        {
                            { "rule_id", fp.RuleId },
                            { "variable", fp.Variable },
                            { "files", fp.Files.Cast<object?>().ToList() },
                        };
                    })
                    .ToList()
            },
        };
        var parent = Path.GetDirectoryName(full);
        if (!string.IsNullOrEmpty(parent))
            Directory.CreateDirectory(parent);
        File.WriteAllText(full, Json.Pretty(baseline) + "\n");
        Console.Error.WriteLine($"info Wrote baseline to {baselinePath}");
    }
}

File: swift/Sources/EnvdoctorKit/Commands/ScanCommand.swift
import Foundation

public enum OutputFormat: String {
    case human, json, sarif
}

public struct ScanArgs {
    public var root: String?
    public var format: OutputFormat = .human
    public var strict = false
    public var verbose = false
    public var only: [String] = []
    public var baseline: String?
    public var writeBaseline: String?
    public var staged = false
    public var since: String?
    /// Write the rendered output to a file instead of stdout (port addition).
    public var output: String?

    public init() {}
}

public enum ScanCommand {
    /// `envdoctor scan` — discover, parse, audit, and report.
    public static func run(_ args: ScanArgs) throws -> Int {
        let root = resolveRoot(args.root)

        // Warn about unknown detector ids passed via --only.
        let known = Audit.allDetectors.map(\.id)
        for rule in args.only where !known.contains(rule) {
            FileHandle.standardError.write("warning Unknown detector \"\(rule)\" (known: \(known.joined(separator: ", ")))\n")
        }

        let gitFilter: GitFilterOptions? = args.staged
            ? GitFilterOptions(staged: true)
            : args.since.map { GitFilterOptions(since: $0) }

        if let gitFilter, Git.hasNoGitChanges(root, gitFilter) {
            print("✓ No changed env-related files to scan")
            return 0
        }

        let context = try Pipeline.loadProject(root, gitFilter: gitFilter)
        var audit = Audit.runAudit(context.model, options: AuditOptions(
            strict: args.strict,
            only: args.only,
            rules: context.config.rules
        ))

        reportParseErrors(context.model, root)

        if let baseline = args.baseline {
            audit = applyBaseline(audit, baseline, root, args.strict)
        }

        if let writeBaseline = args.writeBaseline {
            writeBaselineFile(audit, writeBaseline, root)
        }

        let rendered: String
        switch args.format {
        case .json:
            rendered = ScanJsonFormatter.render(audit, rootDir: root)
        case .sarif:
            rendered = SarifFormatter.renderSarif(audit, rootDir: root)
        case .human:
            rendered = HumanFormatter.renderReport(audit, rootDir: root, verbose: args.verbose)
        }

        if let output = args.output {
            try writeFile(at: output, content: rendered)
        } else {
            print(rendered)
        }

        return audit.exitCode
    }

    // MARK: - Baselines

    private struct BaselineEntry {
        var ruleId: String
        var variable: String
        var files: [String]
    }

    private static func fingerprint(_ root: String, _ finding: Finding) -> BaselineEntry {
        let files = Array(Set(finding.locations.map { displayPath(root, $0.filePath) })).sorted()
        return BaselineEntry(ruleId: finding.ruleId, variable: finding.variable, files: files)
    }

    private static func applyBaseline(_ audit: AuditResult, _ baselinePath: String, _ root: String, _ strict: Bool) -> AuditResult {
        let fullPath = resolvePath(root, baselinePath)
        struct BaselineFile { var version: Int; var findings: [BaselineEntry] }
        let baseline: BaselineFile?
        do {
            let raw = try String(contentsOfFile: fullPath, encoding: .utf8)
            guard case .object(let obj) = try JsonParser.parse(raw) else { throw JsonParser.Error(description: "not an object") }
            var version = 0
            if case .int(let v)? = obj["version"] { version = v }
            var entries: [BaselineEntry] = []
            if case .array(let findings)? = obj["findings"] {
                for f in findings {
                    guard case .object(let fo) = f else { continue }
                    let ruleId = fo["ruleId"]?.asString ?? ""
                    let variable = fo["variable"]?.asString ?? ""
                    let files = fo["files"]?.asArray?.map { $0.asString ?? "" } ?? []
                    entries.append(BaselineEntry(ruleId: ruleId, variable: variable, files: files))
                }
            }
            baseline = BaselineFile(version: version, findings: entries)
        } catch {
            FileHandle.standardError.write("warning Could not read baseline \(baselinePath): \(error.localizedDescription)\n")
            baseline = nil
        }

        guard let baseline else { return audit }

        func matches(_ a: BaselineEntry, _ b: BaselineEntry) -> Bool {
            a.ruleId == b.ruleId && a.variable == b.variable && a.files == b.files
        }

        let before = audit.findings.count
        let findings = audit.findings.filter { f in
            let fp = fingerprint(root, f)
            return !baseline.findings.contains { matches($0, fp) }
        }
        let suppressed = before - findings.count
        if suppressed > 0 {
            FileHandle.standardError.write("info \(suppressed) finding\(suppressed == 1 ? "" : "s") suppressed by baseline\n")
        }

        return recomputeAudit(audit, findings, strict)
    }

    private static func recomputeAudit(_ audit: AuditResult, _ findings: [Finding], _ strict: Bool) -> AuditResult {
        let summary = AuditSummary(
            filesScanned: audit.summary.filesScanned,
            variablesFound: audit.summary.variablesFound,
            errors: findings.filter { $0.severity == .error }.count,
            warnings: findings.filter { $0.severity == .warning }.count,
            infos: findings.filter { $0.severity == .info }.count,
            total: findings.count
        )
        let exitCode = (summary.errors > 0 || (strict && summary.warnings > 0)) ? 1 : 0
        return AuditResult(findings: findings, summary: summary, exitCode: exitCode)
    }

    private static func writeBaselineFile(_ audit: AuditResult, _ baselinePath: String, _ root: String) {
        let fullPath = resolvePath(root, baselinePath)
        var baseline = JSONObject()
        baseline.set("version", .int(1))
        baseline.set("findings", .array(audit.findings.map { f in
            let fp = fingerprint(root, f)
            var entry = JSONObject()
            entry.set("ruleId", .string(fp.ruleId))
            entry.set("variable", .string(fp.variable))
            entry.set("files", .array(fp.files.map { .string($0) }))
            return .object(entry)
        }))
        let dir = (fullPath as NSString).deletingLastPathComponent
        try? FileManager.default.createDirectory(atPath: dir, withIntermediateDirectories: true)
        try? writeFile(at: fullPath, content: Json.pretty(.object(baseline)) + "\n")
        FileHandle.standardError.write("info Wrote baseline to \(baselinePath)\n")
    }
}
